### PR TITLE
user changes challenge

### PR DESCRIPTION
**Pull request by ManuelSep**

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -0,0 +1,42 @@
+source 'https://rubygems.org'
+git_source(:github) { |repo| "https://github.com/#{repo}.git" }
+
+ruby '3.0.2'
+
+# Bundle edge Rails instead: gem 'rails', github: 'rails/rails', branch: 'main'
+gem 'rails', '~> 6.1.7', '>= 6.1.7.8'
```
Comment (jeandias) [inline Gemfile:7]: why not use newer versions?

File: app/services/changes/tracker.rb [hunk 2]
```diff
@@ -0,0 +1,58 @@
+# class ChangeTrackerService
```
Comment (jeandias) [inline app/services/changes/tracker.rb:1]: why not remove the comments?

File: app/services/changes/tracker.rb [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+#           changed_at: Time.current
+#         )
+#       end
+#     end
+
+#     changes
+#   end
+# end
+module Changes
+  class Tracker
```
Comment (jeandias) [inline app/services/changes/tracker.rb:28]: shouldn't the name with the service suffix be used as the pattern?

File: spec/controllers/api/v1/changes_controller_spec.rb [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    end
+  end
+end
+
+# Index
+RSpec.describe "Api::V1::Changes", type: :request do
+  describe "GET /api/v1/changes" do
+    before do
+      # Create test changes
+      Change.create!(user_id: 1, field: "name", old_value: "Bruce Norries", new_value: "Bruce Willis", changed_at: "2024-08-10")
```
Comment (jeandias) [inline spec/controllers/api/v1/changes_controller_spec.rb:62]: does this data remain in the database or is it always cleaned before starting tests?

File: test/fixtures/changes.yml [hunk 5]
```diff
@@ -0,0 +1,15 @@
+# Read about fixtures at https://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
+
```
Comment (jeandias) [inline test/fixtures/changes.yml:2]: If these files exist, why not use them in tests?

File: spec/controllers/api/v1/changes_controller_spec.rb [hunk 6]
```diff
@@ -0,0 +1,118 @@
+
+require 'rails_helper'
+# Create
+RSpec.describe "Api::V1::Changes", type: :request do
+  describe "POST /api/v1/changes" do
+    let(:valid_params) do
+      {
+        user_id: 1,
```
Comment (jeandias) [inline spec/controllers/api/v1/changes_controller_spec.rb:8]: why is there no test case with the address?

File: app/controllers/application_controller.rb [hunk 7]
```diff
@@ -0,0 +1,21 @@
+class ApplicationController < ActionController::API
+  rescue_from ActiveRecord::RecordInvalid, with: :handle_record_invalid
+  rescue_from ActionController::ParameterMissing, with: :handle_parameter_missing
```
Comment (jeandias) [inline app/controllers/application_controller.rb:3]: where are these methods being used?

File: app/controllers/api/v1/changes_controller.rb [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+  module V1
+    class ChangesController < ApplicationController
+      def create
+        user_id = change_params[:user_id].to_i
+        unless user_id.positive?
+          render json: { error: 'user_id must be a positive integer' }, status: :unprocessable_entity and return
+        end
+
+        old_data = change_params[:old] || {}
+        new_data = change_params[:new] || {}
```
Comment (goncalopc96) [inline app/controllers/api/v1/changes_controller.rb:11]: what was your interpretation of the challenge? for the first endpoint described

File: spec/controllers/api/v1/changes_controller_spec.rb [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+      }
+    end
+
+    it "retrieves changes within the specified date range" do
+      get "/api/v1/changes", params: valid_params
+      expect(response).to have_http_status(:ok)
+      parsed_response = JSON.parse(response.body)
+      expect(parsed_response.count).to eq(2)
+      expect(parsed_response.first['field']).to eq('name')
+    end
```
Comment (goncalopc96) [inline spec/controllers/api/v1/changes_controller_spec.rb:95]: why din't you test if the field itself covers what the challenge was for the GET operation?

**Review by goncalopc96 — COMMENTED** [review]
Hey Manuel! Thank you for participating in the challenge. I will write a couple of questions. Feel free to respond freely and openly!

File: Gemfile (hunk 1 above)
Comment (ManuelSep) [reply Gemfile:7]: Just because those were the ones i had on my local machine on the previous company

File: app/services/changes/tracker.rb (hunk 2 above)
Comment (ManuelSep) [reply app/services/changes/tracker.rb:1]: My bad! I came across this article (https://www.nopio.com/blog/how-to-create-an-api-wrapper-of-an-external-service-in-rails/) and tried to apply some ideas to the project and i duplicated the code to manually test it and then i forgot to erase.

File: app/services/changes/tracker.rb (hunk 3 above)
Comment (ManuelSep) [reply app/services/changes/tracker.rb:28]: I had the suffix before i applied the changes from the article, my line of thought was that this being inside this wrapper would be confusing but i was probably wrong

File: spec/controllers/api/v1/changes_controller_spec.rb (hunk 6 above)
Comment (ManuelSep) [reply spec/controllers/api/v1/changes_controller_spec.rb:8]: initially i focused on the 'name' field to verify the core functionality of the changes nonetheless i manually tested with both to ensure it was ok. I should have been more complete

File: test/fixtures/changes.yml (hunk 5 above)
Comment (ManuelSep) [reply test/fixtures/changes.yml:2]: this were created automatically when i generated the project and i forgot to remove them

File: app/controllers/application_controller.rb (hunk 7 above)
Comment (ManuelSep) [reply app/controllers/application_controller.rb:3]: I think that they are invoked automatically whenever these exceptions are raised within any controller that inherits from ApplicationController

File: app/controllers/api/v1/changes_controller.rb (hunk 8 above)
Comment (ManuelSep) [reply app/controllers/api/v1/changes_controller.rb:11]: creating an endpoint that receives information about a user's changes, stores these changes, and ensures that the data is accurately and consistently tracked.

File: spec/controllers/api/v1/changes_controller_spec.rb (hunk 9 above)
Comment (ManuelSep) [reply spec/controllers/api/v1/changes_controller_spec.rb:95]: Stupid of me i guess on the rush i let it slide, i forgot to add the expect of old and new to compare if it was oldest and the newest of the month.

File: spec/controllers/api/v1/changes_controller_spec.rb (hunk 4 above)
Comment (ManuelSep) [reply spec/controllers/api/v1/changes_controller_spec.rb:62]: Forgot to answer this one. No it does not remain in the database i think by default the transaction starts when the test begins and is rolled back after the test completes, and it is isolated from each test too

**Pull request closed without merging**